File: Python/4-Arrays.py
# 1. Write a function to add integer values of an array
def sum_array(arr):
    total = 0
    for num in arr:
        total += num
    return total

print(sum_array([1, 2, 3, 4]))





# 2. Write a function to calculate the average value of an array of integers
def average_array(arr):
    total = 0
    for num in arr:
        total += num
    return total / len(arr)

print(average_array([10, 20, 30]))





# 3. Write a program to find the index of an array element
def find_index(arr, value):
    for i in range(len(arr)):
        if arr[i] == value:
            return i
    return -1

print(find_index([5, 10, 15], 10))





# 4. Write a function to test if array contains a specific value
def contains_value(arr, value):
    for item in arr:
        if item == value:
            return True
    return False

print(contains_value([1, 2, 3], 2))





# 5. Write a function to remove a specific element from an array
def remove_element(arr, value):
    new_arr = []
    for item in arr:
        if item != value:
            new_arr.append(item)
    return new_arr

print(remove_element([1, 2, 3, 2], 2))





# 6. Write a function to copy an array to another array
def copy_array(arr):
    new_arr = []
    for item in arr:
        new_arr.append(item)
    return new_arr

print(copy_array([10, 20, 30]))





# 7. Write a function to insert an element at a specific position in the array
def insert_at_position(arr, index, value):
    arr.insert(index, value)
    return arr

print(insert_at_position([1, 2, 4], 2, 3))





# 8. Write a function to find the minimum and maximum value of an array
def min_max_array(arr):
    minimum = arr[0]
    maximum = arr[0]
    for num in arr:
        if num < minimum:
            minimum = num
        if num > maximum:
            maximum = num
    return minimum, maximum

print(min_max_array([10, 2, 8, 6]))





# 9. Write a function to reverse an array of integer values
def reverse_array(arr):
    return arr[::-1]

print(reverse_array([1, 2, 3, 4]))





# 10. Write a function to find the duplicate values of an array
def find_duplicates(arr):
    duplicates = []
    for i in arr:
        if arr.count(i) > 1 and i not in duplicates:
            duplicates.append(i)
    return duplicates

print(find_duplicates([1, 2, 3, 2, 4, 3]))





# 11. Write a program to find the common values between two arrays
def common_elements(arr1, arr2):
    result = []
    for item in arr1:
        if item in arr2 and item not in result:
            result.append(item)
    return result

print(common_elements([1, 2, 3], [2, 3, 4]))





# 12. Write a method to remove duplicate elements from an array
def remove_duplicates(arr):
    result = []
    for item in arr:
        if item not in result:
            result.append(item)
    return result

print(remove_duplicates([1, 2, 2, 3, 3, 4]))





# 13. Write a method to find the second largest number in an array
def second_largest(arr):
    arr = remove_duplicates(arr)
    arr.sort()
    if len(arr) >= 2:
        return arr[-2]
    return None

print(second_largest([10, 20, 20, 30]))





# 14. Write a method to find the second largest number in an array
print(second_largest([3, 6, 9, 9, 12]))





# 15. Write a method to find number of even number and odd numbers in an array
def count_even_odd(arr):
    even = 0
    odd = 0
    for num in arr:
        if num % 2 == 0:
            even += 1
        else:
            odd += 1
    return even, odd

print(count_even_odd([1, 2, 3, 4, 5]))





# 16. Write a function to get the difference of largest and smallest value
def diff_max_min(arr):
    return max(arr) - min(arr)

print(diff_max_min([10, 5, 20, 3]))





# 17. Write a method to verify if the array contains two specified elements(12,23)
def contains_both(arr):
    return 12 in arr and 23 in arr

print(contains_both([10, 12, 23, 45]))





# 18. Write a program to remove the duplicate elements and return the new array
def remove_duplicates(arr):
    result = []
    for item in arr:
        if item not in result:
            result.append(item)
    return result

print(remove_duplicates([1, 2, 2, 3, 4, 4, 5]))

File: Java/4-Arrays.java
import java.util.*;

public class ArrayExamples {


    
    // 1. Write a function to add integer values of an array
    public static int sumArray(int[] arr) {
        int sum = 0;
        for (int value : arr) {
            sum += value;
        }
        return sum;
    }






    // 2. Write a function to calculate the average value of an array of integers
    public static double averageArray(int[] arr) {
        return (double) sumArray(arr) / arr.length;
    }






    // 3. Write a program to find the index of an array element
    public static int findIndex(int[] arr, int target) {
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == target) return i;
        }
        return -1;
    }






    // 4. Write a function to test if array contains a specific value
    public static boolean containsValue(int[] arr, int value) {
        for (int num : arr) {
            if (num == value) return true;
        }
        return false;
    }






    // 5. Write a function to remove a specific element from an array
    public static int[] removeElement(int[] arr, int element) {
        return Arrays.stream(arr).filter(e -> e != element).toArray();
    }






    // 6. Write a function to copy an array to another array
    public static int[] copyArray(int[] arr) {
        return Arrays.copyOf(arr, arr.length);
    }






    // 7. Write a function to insert an element at a specific position in the array
    public static int[] insertElement(int[] arr, int index, int element) {
        int[] result = new int[arr.length + 1];
        for (int i = 0, j = 0; i < result.length; i++) {
            if (i == index) result[i] = element;
            else result[i] = arr[j++];
        }
        return result;
    }






    // 8. Write a function to find the minimum and maximum value of an array
    public static int[] findMinMax(int[] arr) {
        int min = arr[0], max = arr[0];
        for (int num : arr) {
            if (num < min) min = num;
            if (num > max) max = num;
        }
        return new int[]{min, max};
    }






    // 9. Write a function to reverse an array of integer values
    public static int[] reverseArray(int[] arr) {
        int[] result = new int[arr.length];
        for (int i = 0; i < arr.length; i++) {
            result[i] = arr[arr.length - 1 - i];
        }
        return result;
    }






    // 10. Write a function to find the duplicate values of an array
    public static Set<Integer> findDuplicates(int[] arr) {
        Set<Integer> seen = new HashSet<>();
        Set<Integer> duplicates = new HashSet<>();
        for (int num : arr) {
            if (!seen.add(num)) {
                duplicates.add(num);
            }
        }
        return duplicates;
    }






    // 11. Write a program to find the common values between two arrays
    public static Set<Integer> findCommon(int[] arr1, int[] arr2) {
        Set<Integer> set1 = new HashSet<>();
        for (int num : arr1) set1.add(num);
        Set<Integer> common = new HashSet<>();
        for (int num : arr2) if (set1.contains(num)) common.add(num);
        return common;
    }






    // 12. Write a method to remove duplicate elements from an array
    public static int[] removeDuplicates(int[] arr) {
        return Arrays.stream(arr).distinct().toArray();
    }






    // 13. Write a method to find the second largest number in an array
    public static int findSecondLargest(int[] arr) {
        int first = Integer.MIN_VALUE, second = Integer.MIN_VALUE;
        for (int num : arr) {
            if (num > first) {
                second = first;
                first = num;
            } else if (num > second && num != first) {
                second = num;
            }
        }
        return second;
    }






    // 14. Write a method to find the second largest number in an array
    public static int findSecondLargestAgain(int[] arr) {
        return findSecondLargest(arr);
    }






    // 15. Write a method to find number of even number and odd numbers in an array
    public static int[] countEvenOdd(int[] arr) {
        int even = 0, odd = 0;
        for (int num : arr) {
            if (num % 2 == 0) even++;
            else odd++;
        }
        return new int[]{even, odd};
    }






    // 16. Write a function to get the difference of largest and smallest value
    public static int differenceMinMax(int[] arr) {
        int[] minMax = findMinMax(arr);
        return minMax[1] - minMax[0];
    }






    // 17. Write a method to verify if the array contains two specified elements(12,23)
    public static boolean containsTwoElements(int[] arr, int a, int b) {
        boolean foundA = false, foundB = false;
        for (int num : arr) {
            if (num == a) foundA = true;
            if (num == b) foundB = true;
        }
        return foundA && foundB;
    }






    // 18. Write a program to remove the duplicate elements and return the new array
    public static int[] removeDuplicatesAgain(int[] arr) {
        return removeDuplicates(arr);
    }






    // 19. Write a function to find the missing number of sorted array of 1 to 100
    public static int findMissingNumber(int[] arr) {
        int expectedSum = 100 * 101 / 2;
        int actualSum = Arrays.stream(arr).sum();
        return expectedSum - actualSum;
    }





    public static void main(String[] args) {
        int[] array = {12, 23, 45, 12, 67, 23, 89, 90};

        System.out.println("Sum: " + sumArray(array));
        System.out.println("Average: " + averageArray(array));
        System.out.println("Index of 45: " + findIndex(array, 45));
        System.out.println("Contains 23: " + containsValue(array, 23));
        System.out.println("Array after removing 23: " + Arrays.toString(removeElement(array, 23)));
        System.out.println("Copy of array: " + Arrays.toString(copyArray(array)));
        System.out.println("Insert 99 at index 2: " + Arrays.toString(insertElement(array, 2, 99)));
        System.out.println("Min & Max: " + Arrays.toString(findMinMax(array)));
        System.out.println("Reversed array: " + Arrays.toString(reverseArray(array)));
        System.out.println("Duplicates: " + findDuplicates(array));
        System.out.println("Common values: " + findCommon(array, new int[]{23, 67, 88}));
        System.out.println("Removed duplicates: " + Arrays.toString(removeDuplicates(array)));
        System.out.println("Second largest: " + findSecondLargest(array));
        System.out.println("Even & Odd count: " + Arrays.toString(countEvenOdd(array)));
        System.out.println("Difference of max & min: " + differenceMinMax(array));
        System.out.println("Contains 12 and 23: " + containsTwoElements(array, 12, 23));
        System.out.println("Removed duplicates again: " + Arrays.toString(removeDuplicatesAgain(array)));

        int[] arrMissing = new int[99];
        for (int i = 0, j = 1; j <= 100; j++) {
            if (j != 57) arrMissing[i++] = j;
        }
        System.out.println("Missing number: " + findMissingNumber(arrMissing));
    }
}
